src/reaktoro_pse/tutorials/softening_acid_ro_example.ipynb

```diff
@@ -0,0 +1,1324 @@
+    "m.fs.softening.effluent_pH = Var(initialize=7,bounds=(5, 12), units=pyunits.dimensionless)\n",
+    "m.fs.softening.effluent_alkalinity = Var(\n",
+    "    initialize=100, units=pyunits.mg / pyunits.L, bounds=(30, None)\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Mass balance constraints for softening process\n",
+    "\n",
+    "For each ion where reagent and solid is participating \n",
+    "\n",
+    "$$ \\dot{m}_{ion,out}= \\dot{m}_{ion,in} + \\dot{m}_{ion,reagent} -\\dot{m}_{ion,solid} $$\n",
+    "\n",
+    "Example for Calcium:\n",
+    "$$ \\dot{m}_{Ca,out}= \\dot{m}_{Ca,in} + \\dot{m}_{Ca,CaO} -\\dot{m}_{Ca,Calcite} $$\n",
+    "\n",
+    "Otherwise:\n",
+    "\n",
+    "$$ \\dot{m}_{ion,out}= \\dot{m}_{ion,in} $$"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "m.fs.softening.effluent_species_mass_flows = Var(\n",
+    "    ions, initialize=1e-5, units=pyunits.kg / pyunits.s\n",
+    ")\n",
+    "\n",
+    "\n",
+    "@m.fs.softening.Constraint(ions)\n",
+    "def eq_effluent_species_mass_flows(fs, ion):\n",
+    "    if \"Ca\" == ion:\n",
+    "        return m.fs.softening.effluent_species_mass_flows[\"Ca\"] == (\n",
+    "            m.fs.softening.reagents_mol_flow[\"CaO\"] * m.fs.feed.specie_mw[\"Ca\"]\n",
+    "            + m.fs.feed.species_mass_flow[\"Ca\"]\n",
+    "            - m.fs.softening.precipitants[\"Calcite\"]\n",
+    "            * m.fs.feed.specie_mw[\"Ca\"]\n",
+    "        )\n",
+    "    elif \"Na\" == ion:\n",
+    "        return m.fs.softening.effluent_species_mass_flows[\"Na\"] == (\n",
+    "            2\n",
+    "            * m.fs.softening.reagents_mol_flow[\"Na2CO3\"]\n",
+    "            * m.fs.feed.specie_mw[\"Na\"]\n",
+    "            + m.fs.feed.species_mass_flow[\"Na\"]\n",
+    "        )\n",
+    "    elif \"HCO3\" == ion:\n",
+    "        return m.fs.softening.effluent_species_mass_flows[\"HCO3\"] == (\n",
+    "            m.fs.feed.species_mass_flow[\"HCO3\"]\n",
+    "            + m.fs.softening.reagents_mol_flow[\"Na2CO3\"]\n",
+    "            * m.fs.feed.specie_mw[\"HCO3\"]\n",
+    "            - m.fs.softening.precipitants[\"Calcite\"]\n",
+    "            * m.fs.feed.specie_mw[\"HCO3\"]\n",
+    "        )\n",
+    "    else:\n",
+    "        return (\n",
+    "            m.fs.softening.effluent_species_mass_flows[ion]\n",
+    "            == m.fs.feed.species_mass_flow[ion]\n",
+    "        )\n",
+    "\n",
+    "@m.fs.softening.Constraint([\"CaO\", \"Na2CO3\"])\n",
+    "def eq_dose(fs, reagent):\n",
+    "    return m.fs.softening.reagent_dose[reagent] == pyunits.convert(\n",
+    "        m.fs.softening.reagents_mol_flow[reagent]\n",
+    "        * m.fs.softening.reagents_mw[reagent]\n",
+    "        / m.fs.feed.species_mass_flow[\"H2O\"],\n",
+    "        to_units=pyunits.mg / pyunits.kg,\n",
+    "    )"
+   ]
+  },
```

Comment:
Same idea here-- aggregate all the cells that do the work of setting up the softening model. You can outline the main operational steps that you want to point out, below the suggested "Step2" markdown. 

I'd say if there is a REALLY interesting segment that you want to isolate because it is just so compelling, then that would be reason enough to do so. Otherwise, I think we should aggregate. 

We're working our way to the first punchline: estimating scaling tendency and pH